storage/redis/peer_store.go

```diff
@@ -0,0 +1,762 @@
+				numSeeders++
+				infohashesMap[ih] = struct{}{}
+			} else {
+				log.Error("storage: populateProm invalid Redis state", log.Fields{
+					"Hkey":   group,
+					"Hfield": ih,
+				})
+			}
+		}
+
+		numInfohashes += uint64(len(infohashesMap))
+	}
+
+	storage.PromInfohashesCount.Set(float64(numInfohashes))
+	storage.PromSeedersCount.Set(float64(numSeeders))
+	storage.PromLeechersCount.Set(float64(numLeechers))
+}
+
+// recordGCDuration records the duration of a GC sweep.
+func recordGCDuration(duration time.Duration) {
+	log.Debug("storage: recordGCDuration", log.Fields{"timeTaken(ms)": float64(duration.Nanoseconds()) / float64(time.Millisecond)})
+	storage.PromGCDurationMilliseconds.Observe(float64(duration.Nanoseconds()) / float64(time.Millisecond))
+}
+
+func (ps *peerStore) getClock() int64 {
+	return timecache.NowUnixNano()
+}
+
+func (ps *peerStore) PutSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	addressFamily := p.IP.AddressFamily.String()
+	log.Debug("storage: PutSeeder", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     p.String(),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	// Update the peer in the swarm.
+	encodedSeederInfoHash := addressFamily + "_S_" + ih.String()
+	ct := ps.getClock()
+	_, err := conn.Do("HSET", encodedSeederInfoHash, pk, ct)
+	if err != nil {
+		return err
+	}
+	_, err = conn.Do("HSET", addressFamily, encodedSeederInfoHash, ct)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ps *peerStore) DeleteSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	addressFamily := p.IP.AddressFamily.String()
+	log.Debug("storage: DeleteSeeder", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     p.String(),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	encodedSeederInfoHash := addressFamily + "_S_" + ih.String()
+
+	delNum, err := conn.Do("HDEL", encodedSeederInfoHash, pk)
+	if err != nil {
+		return err
+	}
+	if delNum.(int64) == 0 {
+		return storage.ErrResourceDoesNotExist
+	}
+
+	return nil
+}
+
+func (ps *peerStore) PutLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	addressFamily := p.IP.AddressFamily.String()
+	log.Debug("storage: PutLeecher", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     p.String(),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	// Update the peer in the swarm.
+	encodedLeecherInfoHash := addressFamily + "_L_" + ih.String()
+	ct := ps.getClock()
+	_, err := conn.Do("HSET", encodedLeecherInfoHash, pk, ct)
+	if err != nil {
+		return err
+	}
+	_, err = conn.Do("HSET", addressFamily, encodedLeecherInfoHash, ct)
```

Comment:
Why not do both of these in a transaction using a lua script or `MULTI`?

Reply:
good point